lib/routes/account.js

```diff
@@ -1021,50 +1048,78 @@ module.exports = function (
-        db.account(Buffer(uid, 'hex'))
-          .then(
-            function (account) {
-              // If the account is already verified, they may be e.g.
-              // clicking a stale link.  Silently succeed.
-              if (account.emailVerified) {
-                log.increment('account.already_verified')
-                return true
-              }
-              if (!butil.buffersAreEqual(code, account.emailCode)) {
-                throw error.invalidVerificationCode()
-              }
-              log.timing('account.verified', Date.now() - account.createdAt)
-              log.event('verified', { email: account.email, uid: account.uid, locale: account.locale })
-              log.increment('account.verified')
 
-              // send a push notification to all devices that the account changed
-              push.notifyUpdate(uid)
+        if(code.length === 32) {
+          // Hex code length of 32 signifies a tokenId, perform login token verification
+          log.begin('Account.RecoveryLoginVerify', request);
+          db.verifyToken(code, {uid: uid})
+            .then(function () {
+            })
+            .done(
+              function () {
+                reply({})
+              },
+              reply
+            )
+        } else {
+          log.begin('Account.RecoveryEmailVerify', request)
+          db.account(Buffer(uid, 'hex'))
+            .then(
+              function (account) {
+                // If the account is already verified, they may be e.g.
+                // clicking a stale link.  Silently succeed.
+                if (account.emailVerified) {
+                  log.increment('account.already_verified')
+                  return true
+                }
+                if (!butil.buffersAreEqual(code, account.emailCode)) {
+                  throw error.invalidVerificationCode()
+                }
+                log.timing('account.verified', Date.now() - account.createdAt)
+                log.event('verified', {email: account.email, uid: account.uid, locale: account.locale})
+                log.increment('account.verified')
 
-              return db.verifyEmail(account)
-                .then(mailer.sendPostVerifyEmail.bind(
-                    mailer,
-                    account.email,
-                    {
-                      acceptLanguage: request.app.acceptLanguage
-                    }
-                  )
-                )
-            }
-          )
-          .done(
-            function () {
-              reply({})
-            },
-            reply
-          )
+                // send a push notification to all devices that the account changed
+                push.notifyUpdate(uid)
+
+                return db.sessions(account.uid)
+                  .then(function(sessions) {
+                    // TODO How to know which session is associated with the email verification?
+                    // Assuming last created session?
```

Comment:
Actually, thinking about the edge-cases here more, I think we will need to allow *any* valid `tokenVerificationId` to flip the account into the `verified` state.  Otherwise, consider the following scenario:

* I create an account on one device, but never get around to verifying it.  I lose the email and the original sessionToken.
* Some time later, I sign in on a second device.  It sends me a "confirm this sign-in" email and I click through it
* I now have an unverified account with a verified sessionToken.

Given that we're re-using the email verification endpoints, I fear such a state could only result in badness.  Depending on how we choose to modify the `recovery_email/resend_code` endpoint, it's not obvious that I'd ever be able to move my account into the verified state from this point.

A suggestion (and not a completely-thought-out one...) for the logic we may want here is something along the lines of:

* Verify any tokens matching the submitted code
* If some tokens were matched, check whether the account is unverified and mark it verified if so
* If no tokens were matched, check whether it matches the account verification code, and verify the account if so
* If the code didn't match in either place, throw an `invalidVerificationCode` error

That's a bit yuck, but I suspect it's the price we pay for overloading these endpoints.